src/CarbonAware.WebApi/test/integrationTests/_snapshots/CarbonAwareControllerTests_BestLocations_ReturnsOK.json

```diff
@@ -0,0 +1,20 @@
+{
```

Comment:
I'm sorry, but still I don't get it with files why are here based on your comment about this are autogenerated by the new proposed package. (BTW, I think if there is a new package coming in, it seems to be approved to some degree, but better to ask first before committing to it)

Reply:
I think this doc explains it best. https://theramis.github.io/Snapper/#/?id=what-is-snapper

Reply:
> I think this doc explains it best. https://theramis.github.io/Snapper/#/?id=what-is-snapper

@in4margaret , thank you for the extra information, definitely something to look for.
I think for this integration tests specifically, it is not bringing any value since the output is the same as the one that is generated. Have you look how the data is generated for the integration testing path? Unfortunately
it is **static**, so  basically the tests that are added with this PR never are going to fail.
I was hoping to have better generation of content for this PR for the integration.

Reply:
Snapshot testing alone has uncovered multiple problems with the existing code:

1. One of the endpoint didn't have correct query parameter mappings and calling it parameters passed in url resulted in OK 200 response even though parameters weren't applied. This means that testing for 200 OK response alone would not cover this case. And if we need to test part of the response for correctness we might as well test the whole thing with a snapshot. Note that url query parameters were fixed in this PR.  
![image](https://user-images.githubusercontent.com/5702241/222495572-61d0a1c3-28b6-40b4-af87-3117defe68e6.png)
2. Integration test weren't actually using the correct json file for the integration tests. I was able to find this problem only because snapshot testing was added and I've noticed that returning data was not from the mock json datasource.
![image](https://user-images.githubusercontent.com/5702241/222497165-74dd1798-7115-4043-9d6c-fe4bf2fffedc.png)
3. I've noticed from the snapshot testing that the dates format (timezone) that we return was dependent on the input of the user, output of the datasource and local server timezone when we return it in Web API. I've only cought it because snapshot testing was introduced and the tests started failing in the pipelines where the timezone is different from my local one.
4. From examining the snapshot output I've found that some of the WebAPI endpoints don't return DTO models (they return GSF models) even though they should. 

Having tests that test only OK 200 response would not have cached this problems, So I think that this approach is better to have than not.

Reply:
> Snapshot testing alone has uncovered multiple problems with the existing code:
> 
> 1. One of the endpoint didn't have correct query parameter mappings and calling it parameters passed in url resulted in OK 200 response even though parameters weren't applied. This means that testing for 200 OK response alone would not cover this case. And if we need to test part of the response for correctness we might as well test the whole thing with a snapshot. Note that url query parameters were fixed in this PR.
>    ![image](https://user-images.githubusercontent.com/5702241/222495572-61d0a1c3-28b6-40b4-af87-3117defe68e6.png)
> 2. Integration test weren't actually using the correct json file for the integration tests. I was able to find this problem only because snapshot testing was added and I've noticed that returning data was not from the mock json datasource.
>    ![image](https://user-images.githubusercontent.com/5702241/222497165-74dd1798-7115-4043-9d6c-fe4bf2fffedc.png)
> 3. I've noticed from the snapshot testing that the dates format (timezone) that we return was dependent on the input of the user, output of the datasource and local server timezone when we return it in Web API. I've only cought it because snapshot testing was introduced and the tests started failing in the pipelines where the timezone is different from my local one.
> 4. From examining the snapshot output I've found that some of the WebAPI endpoints don't return DTO models (they return GSF models) even though they should.
> 
> Having tests that test only OK 200 response would not have cached this problems, So I think that this approach is better to have than not.

Thank you for the details here.
1. I don't think this is related with having the snapshot nugget package. Am I right?
2. `demo.json` is fine. Please look at : https://github.com/Green-Software-Foundation/carbon-aware-sdk/blob/343113b2a563a021b8f92b3809029e879621737c/src/CarbonAware.WebApi/test/integrationTests/CarbonAware.WebApi.IntegrationTests.csproj#L35

Still the usage of the snapshot is not bringing value imo. It is adding too much burden to the code and hard to maintain in the long run for this kind of functionality.

Reply:
2. If we want to use `demo.json` instead of JSON datasource mocker we should delete the code from IntegrationTestingBase.cs that creates JSON datasource mocker `_dataSourceMocker = new JsonDataSourceMocker();` and delete code/refactor code from Integration tests that calls the data source mocker for JSON data source to generate tests data. Otherwise we have a bunch of code that does nothing but increases test execution time.

Reply:
> 2\. If we want to use `demo.json` instead of JSON datasource mocker we should delete the code from IntegrationTestingBase.cs that creates JSON datasource mocker `_dataSourceMocker = new JsonDataSourceMocker();` and delete code/refactor code from Integration tests that calls the data source mocker for JSON data source to generate tests data. Otherwise we have a bunch of code that does nothing but increases test execution time.

Hi @in4margaret, what you described, should be part of story [258](https://github.com/microsoft/carbon-aware-sdk/issues/258). The reason behind that 'new'  json mocker instance was a [bug](https://github.com/Green-Software-Foundation/carbon-aware-sdk/commit/aeac19610a9c156c213c2171fb8922f500220f68) found based on the dates that expired from the original json file. Hopefully issue 258 would make this better.